internal/storage/storage.go

```diff
@@ -0,0 +1,115 @@
+package storage
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+)
+
+func ArchiveDir(dir string) (out string, err error) {
+	pwd, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		if deferErr := os.Chdir(pwd); deferErr != nil && err == nil {
+			err = deferErr
+		}
+	}()
+
+	if err := os.Chdir(filepath.Dir(dir)); err != nil {
+		return "", err
+	}
+
+	dir = "./"
+	if f, err := os.Stat(dir); os.IsNotExist(err) || !f.IsDir() {
+		return "nil", fmt.Errorf("invalid dir path: %s", dir)
+	}
+
+	tf, err := os.CreateTemp("", "tf-")
+	if err != nil {
+		return "", err
+	}
+	tmpName := tf.Name()
+	defer func() {
+		if err != nil {
+			os.Remove(tmpName)
+		}
+	}()
+
+	h := sha1.New()
+	mw := io.MultiWriter(h, tf)
+
+	gw := gzip.NewWriter(mw)
+	tw := tar.NewWriter(gw)
+	if err := filepath.Walk(dir, func(p string, fi os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Ignore anything that is not a file (directories, symlinks)
+		if !fi.Mode().IsRegular() {
+			return nil
+		}
+
+		header, err := tar.FileInfoHeader(fi, p)
+		if err != nil {
+			return err
+		}
+		// The name needs to be modified to maintain directory structure
+		// as tar.FileInfoHeader only has access to the base name of the file.
+		// Ref: https://golang.org/src/archive/tar/common.go?#L626
+		relFilePath := p
+		if filepath.IsAbs(dir) {
+			relFilePath, err = filepath.Rel(dir, p)
+			if err != nil {
+				return err
+			}
+		}
+		header.Name = relFilePath
+
+		if err := tw.WriteHeader(header); err != nil {
+			return err
+		}
+
+		f, err := os.Open(p)
+		if err != nil {
+			f.Close()
+			return err
+		}
+		if _, err := io.Copy(tw, f); err != nil {
+			f.Close()
+			return err
+		}
+		return f.Close()
+	}); err != nil {
+		tw.Close()
+		gw.Close()
+		tf.Close()
```

Comment:
I think we can put these `Close` calls into a defer, so we don't have to call `x.Close()` for all `if err := ..` block in the next few lines. Go doesn't cry about closing them multiple times.

```go
❯ cat main.go
package main

import (
	"compress/gzip"
	"crypto/sha1"
	"io"
	"log"
	"os"
)
❯ go run .


func main() {
	tf, err := os.CreateTemp("", "tf-")
	if err != nil {
		log.Fatal(err)
	}
	tmpName := tf.Name()
	defer func() {
		if err != nil {
			os.Remove(tmpName)
		}
	}()

	h := sha1.New()
	mw := io.MultiWriter(h, tf)
	gw := gzip.NewWriter(mw)

	gw.Close()
	gw.Close()
	gw.Close()
	gw.Close()
	gw.Close()
	log.Println("Done :)")
}
```

```
❯ go run .
2023/11/30 15:37:01 Done :)
```